openmls/src/treesync/node/leaf_node.rs

```diff
@@ -263,53 +285,122 @@ impl LeafNode {
+        &self,
+        version: ProtocolVersion,
+        ciphersuite: Ciphersuite,
+        required_capabilities: impl Into<Option<&'a RequiredCapabilitiesExtension>>,
+        signature_keys: &[SignaturePublicKey],
+        encryption_keys: &[EncryptionKey],
+        members_supported_credentials: &[&[CredentialType]],
+        currently_in_use: &[CredentialType],
+    ) -> Result<&Self, LeafNodeValidationError> {
+        self.validate(
+            version,
+            ciphersuite,
+            required_capabilities,
+            signature_keys,
+            encryption_keys,
+            members_supported_credentials,
+            currently_in_use,
+        )?;
 
+        // If the LeafNode appears in the leaf_node value of the UpdatePath in a Commit, verify
+        // that leaf_node_source is set to commit.
         match self.payload.leaf_node_source {
             LeafNodeSource::Commit(_) => Ok(self),
             _ => Err(LeafNodeValidationError::InvalidLeafNodeSource),
         }
     }
 
     /// Basic validation of leaf node called in all `validate_in_*` methods.
-    // TODO(#1186)
-    #[allow(unused)]
-    fn validate<'a>(
+    pub(crate) fn validate<'a>(
         &self,
+        version: ProtocolVersion,
+        ciphersuite: Ciphersuite,
         required_capabilities: impl Into<Option<&'a RequiredCapabilitiesExtension>>,
         signature_keys: &[SignaturePublicKey],
         encryption_keys: &[EncryptionKey],
         members_supported_credentials: &[&[CredentialType]],
         currently_in_use: &[CredentialType],
     ) -> Result<&Self, LeafNodeValidationError> {
-        self.validate_required_capabilities(required_capabilities)?
-            .validate_that_capabilities_contain_extension_types()?
+        // The client verifies the validity of a LeafNode using the following steps:
+        self
+            // Verify that the credential in the LeafNode is valid as described in Section 5.3.1.
+            // ...
+            // Verify that the signature on the LeafNode is valid using signature_key.
+            // ...
+            // Verify that the LeafNode is compatible with the group's parameters.
+            // (ValSem106)
+            .validate_version_ciphersuite(version, ciphersuite)?
+            // If the GroupContext has a required_capabilities extension, then the required extensions, proposals, and credential types MUST be listed in the LeafNode's capabilities field.
+            // (ValSem106)
+            .validate_required_capabilities(required_capabilities)?
             .validate_that_capabilities_contain_credential_type()?
-            .validate_that_signature_key_is_unique(signature_keys)?
-            .validate_that_encryption_key_is_unique(encryption_keys)?
+            // Verify that the credential type is supported by all members of the group, as specified by the capabilities field of each member's LeafNode, ...
             .validate_against_group_credentials(members_supported_credentials)?
```

Comment:
This function name seems a little off. We use it to verify member credentials, but the leaf node doesn't know anything about other members. So what it's really doing is verifying that all the credential types it gets in the list are supported by this leaf node.